[fixes #12] prune stale cache entries

stefanpenner opened the pull request:
- [x] get feedback
- [x] write tests to ensure rebuild actually prunes files (will follow up this afternoon)

stefanpenner committed:
[fixes #12] prune stale cache entries

stefanpenner commented on filter.js:
```diff
@@ -43,6 +45,11 @@ Filter.prototype.rebuild = function() {
   var paths = walkSync(srcDir);
+
+  this._cache.deleteExcept(paths).forEach(function(relativePath) {
+    rimraf(path.join(destDir, relativePath));
```
I need to add a test for this code-path yet

caitp commented:
the approach looks okay --- I would be fine with moving the cache.js stuff into filter.js just to keep it small, (in terms of file count) but I'll leave that up to you. I feel like it's something @joliss might want to move to a more general utility module or something.

Lets just get that extra line covered and it should be good to land

stefanpenner commented:
@caitp im actually going to advocate we keep the seperation. It has several benefits.
1. easier to digest SRPP
2. enables the cache to be a pluggable strategy (making something like persistent-filter merely a different cache strategy.
3. While doing this, I actually realized the caching code is overly complicated. Next commit will come with something dramatically reduced and clearer.

caitp commented:
Ok, if you prefer keeping it separate then I'm fine with that, but for small modules I do like keeping them to a single file

stefanpenner commented:
another NPM module it is then I suppose.

caitp commented:
Like I said, leaving it up to you whether to separate it or not. It can evolve into a separate module if it needs to

stefanpenner commented:
I'll likely make it then following the API of https://github.com/stefanpenner/sync-disk-cache, that will make it a quick replacement for a persistent filter variant.

stefanpenner commented:
finally had a chance to finish up the tests. My current status is, getting trolled by mock-fs, it appears to think a file is a folder.

Im digging in further, to figure out how this is happening.

caitp commented:
I'll take a look

stefanpenner commented:
@caitp no need, it was user-error :P

![](http://i.imgur.com/xVyoSl.jpg)

caitp commented:
![qemcogg](https://cloud.githubusercontent.com/assets/2294695/8731331/a7a4893a-2bc4-11e5-94be-ce1d9cc3e020.jpg)

Code generally looks good, once travis is green it can probably be closed

stefanpenner committed:
simplify caching layer, which enables simpler cache eviction

It appears that the caching was written to handle scenarios this filter never uses.
An example of this would be that a cacheInfo supported multiple inputFiles.

stefanpenner committed:
upgrade mock-fs

stefanpenner committed:
add some use-full package.json scripts (for debugging tests)

stefanpenner committed:
add IO.js to travis

stefanpenner committed:
move coverage to npm run script

stefanpenner commented:
I always forget when we need path.join and when we don't. And as this repo doesn't currently have windows CI, we should likely confirm this actually works on windows.

~~I'll add appveyor shortly~~ actualy @caitp are you able to add it to your account (that way you are able to administer it etc) ? I'll glady get it all working though.

stefanpenner committed:
add appveyor

stefanpenner commented:
lets see what our new windows Ci thinks: https://ci.appveyor.com/project/embercli/cauliflower-filter

caitp commented:
added for my account, just need to disable the appvery spammail

stefanpenner commented:
appveyor appears green. (i did disable coverage on appveyor, but i suspect we don't need coverage on both appveyor and travis)

stefanpenner commented:
Feel free to merge this, but I intend to test this out in some real apps after lunch. To confirm the test coverage is sufficient. So merge at your own risk.

stefanpenner commented:
Also if time permits, there may be a follow up PR to adjust the caching layer slightly to be more conducive to a being pluggable (but i didn't want to block this PR on that)

caitp commented:
I'll give you some time to try it out. We aren't using this in angular2, so I don't have any app code that needs it atm

stefanpenner commented:
just gave this a whirl in 2 apps, all appears well.

If issues arise, cc me on them. I will try to fix ASAP.

caitp commented on filter.js:
```diff
@@ -112,45 +117,25 @@ Filter.prototype.processAndCacheFile =
-      mkdirp.sync(path.dirname(dest));
-      symlinkOrCopySync(self.cachePath + '/' + cacheFiles[i], dest);
+    if (fs.existsSync(entry.cacheFile)) {
+      fs.unlinkSync(entry.cacheFile);
```
This line never seems to get hit --- I'm guessing because of the `deleteExcept()` method. I'm also not seeing any lines from coverage.js in coveralls, so it's hard to see if `deleteExcept()` is the cause of this not getting hit or not.

If the line isn't used, lets get rid of it, and if it is used, lets make sure coverage reports make this clear

stefanpenner replied:
it is used, when the cache entry is stale and must be replaced. I'll add this scenario to the suite.

caitp commented:
other than that last bit, lgtm

stefanpenner committed:
add test to ensure stale entries are correctly replaced

caitp commented:
Great stuff, thanks
